volatility3/framework/plugins/windows/amcache.py

```diff
@@ -0,0 +1,603 @@
+# This file is Copyright 2019 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import datetime
+import enum
+import logging
+from itertools import groupby
+from typing import Any, Dict, Iterator, List, NamedTuple, Optional, Tuple, Union
+
+from volatility3.framework import exceptions, interfaces, renderers
+from volatility3.framework.configuration import requirements
+from volatility3.framework.layers import registry
+from volatility3.framework.renderers import conversion
+from volatility3.framework.symbols.windows.extensions import registry as reg_extensions
+from volatility3.plugins import timeliner
+from volatility3.plugins.windows.registry import hivelist
+
+vollog = logging.getLogger(__name__)
+
+
+class AmcacheEntryType(enum.IntEnum):
+    Driver = 1
+    Program = 2
+    File = 3
+
+
+AmcacheEntry = NamedTuple(
+    "AmcacheEntry",
+    [
+        ("Path", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("Company", Union[str, interfaces.renderers.BaseAbsentValue]),
+        (
+            "LastModifyTime",
+            Union[datetime.datetime, interfaces.renderers.BaseAbsentValue],
+        ),
+        (
+            "LastModifyTime2",
+            Union[datetime.datetime, interfaces.renderers.BaseAbsentValue],
+        ),
+        ("InstallTime", Union[datetime.datetime, interfaces.renderers.BaseAbsentValue]),
+        ("CompileTime", Union[datetime.datetime, interfaces.renderers.BaseAbsentValue]),
+        ("SHA1", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("Service", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("ProductName", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("EntryType", str),
+    ],
+)
+
+
+def _try_get_value(
+    values: List[reg_extensions.CM_KEY_VALUE],
+    name: str,
+    expected_type: reg_extensions.RegValueTypes,
+) -> Any:
+    try:
+        value = next(
+            (
+                val.decode_data()
+                for val in values
+                if val.get_name().lower().strip().strip("\u0000") == name.lower()
+            ),
+            None,
+        )
+    except exceptions.InvalidAddressException:
+        return renderers.UnreadableValue()
+    if value is None:
+        return renderers.NotAvailableValue()
+
+    if expected_type == reg_extensions.RegValueTypes.REG_SZ:
+        if isinstance(value, bytes):
+            return value.decode("utf-16le", errors="replace").rstrip("\u0000")
+        else:
+            return renderers.UnparsableValue()
+
+    if expected_type in [
+        reg_extensions.RegValueTypes.REG_DWORD,
+        reg_extensions.RegValueTypes.REG_QWORD,
+        reg_extensions.RegValueTypes.REG_DWORD_BIG_ENDIAN,
+    ]:
+        if isinstance(value, int):
+            return value
+        else:
+            return renderers.UnparsableValue()
+
+    return value
+
+
+def _sort_entries(entry_tuple):
```

Comment:
Feels as though this could be a lambda, but again not a show stopper, and it offers clarity, but if it's gonna be a new function then proper typing (and mabye even a docstring) please.  Just because it's short doesn't mean it needs no explaining...

Reply:
I split it out into a new function since it's essential that the implementations match for `group_by` and `sorted`, added a comment to it plus type hints.